internal/hcsoci/network.go

```diff
@@ -68,3 +70,62 @@ func GetNamespaceEndpoints(ctx context.Context, netNS string) ([]*hns.HNSEndpoin
 	}
 	return endpoints, nil
 }
+
+// NetworkSetup is used to abstract away the details of setting up networking
+// for a container.
+type NetworkSetup interface {
+	ConfigureNetworking(ctx context.Context, namespaceID string) error
+}
```

Comment:
After thinking about this more, I really feel like this is unnecessary. The differentiation of `LocalNetworkSetup` and `ExternalNetworkSetup` is only used to distinguish between if ncproxy is enabled or not. But at every point that we setup this `NetworkSetup` whether it's local or external and then later run `ConfigureNetworking` on that struct, we have access to the parent and thus to information on if ncproxy is enabled. I think it would be better to instead have something like 
```
type NetworkSetupConfig struct {
    VM *uvm.UtilityVM
    CaAddr string 
    ContainerID string
}
```

Then a function like:
```
func (n *NetworkSetupConfig) ConfigureNetworking(ctx context.Context, namespaceID string) error {
    if n.VM.NCProxyEnabled() {
        return n.configureExternalNetworking 
    }
    return n.configureInternalNetworking 
}
```

I feel like this would be cleaner.

Reply:
@ambarve Any thoughts here? Reason I have it as an interface currently is it was easy to kind of swap in different network setup implementations at will to try out certain things and it seems more flexible for future cases if we ever arrive at a third, fourth, etc. way to setup networking.

Reply:
@ambarve Ping again as this is likely buried in your emails. This was the one we were asking about in standup 😆

Reply:
Looking at the this PR, hopefully will get to this soon.

Reply:
Is the question here basically about having an interface with two implementations, where one is chosen based on if ncproxy is being used, versus simply calling a function that chooses one of two other functions to call based on the same condition?

If so, I don't have a super strong opinion either way. If we go one way and it becomes desirable to change in the future, that should be easy to do.

One nice feature of sticking with the existing approach is we can move the code which connects to ncproxy into a `NewExternalNetworkSetup` function, instead of having it in `pod.go`.

Reply:
@kevpar Yep you're spot on

Reply:
I think for now I'll leave it as is unless we run into anything

Reply:
Based on Kevin's other comment about moving this to the `uvm` package, I feel like it would make sense to make these functions off of the `uvm` struct. It could remove the need to pass around this `netSetup` struct around as well.

Reply:
I left it as is and added it to the Resources struct for cleanup. There's a new third parameter now that matches what they want to do for cleanup.

```go
// NetworkConfigType specifies the action to be performed during network configuration.
// For example: setup or teardown
type NetworkConfigType uint8

const (
	NetworkRequestSetup NetworkConfigType = iota
	NetworkRequestTearDown
)

// NetworkSetup is used to abstract away the details of setting up networking
// for a container.
type NetworkSetup interface {
	ConfigureNetworking(ctx context.Context, namespaceID string, configType NetworkConfigType) error
} 
```

where on cleanup in `ReleaseResources`, `ConfigureNetworking` would be called once again and simply passed the teardown request instead.